retrieval/manager.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2013 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package retrieval
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// Appendable returns an Appender.
+type Appendable interface {
+	Appender() (storage.Appender, error)
+}
+
+// NewScrapeManager is the ScrapeManager constructor
+func NewScrapeManager(logger log.Logger, app Appendable) *ScrapeManager {
+
+	return &ScrapeManager{
+		append:        app,
+		logger:        logger,
+		actionCh:      make(chan func()),
+		scrapeConfigs: make(map[string]*config.ScrapeConfig),
+		scrapePools:   make(map[string]*scrapePool),
+		graceShut:     make(chan struct{}),
+	}
+}
+
+// ScrapeManager maintains a set of scrape pools and manages start/stop cicles
+// when receiving new target groups form the discovery manager.
+type ScrapeManager struct {
+	logger        log.Logger
+	append        Appendable
+	scrapeConfigs map[string]*config.ScrapeConfig
+	scrapePools   map[string]*scrapePool
+	actionCh      chan func()
+	graceShut     chan struct{}
+}
+
+// Run starts background processing to handle target updates and reload the scraping loops.
+func (m *ScrapeManager) Run(tsets <-chan map[string][]*config.TargetGroup) error {
+	level.Info(m.logger).Log("msg", "Starting scrape manager...")
+
+	for {
+		select {
+		case f := <-m.actionCh:
+			f()
+		case ts := <-tsets:
+			m.reload(ts)
+		case <-m.graceShut:
+			return nil
+		}
+	}
+}
+
+// Stop cancels all running scrape pools and blocks until all have exited.
+func (m *ScrapeManager) Stop() {
+	for _, sp := range m.scrapePools {
+		sp.stop()
+	}
+	close(m.graceShut)
+}
+
+// ApplyConfig resets the manager's target providers and job configurations as defined by the new cfg.
+func (m *ScrapeManager) ApplyConfig(cfg *config.Config) error {
+	done := make(chan struct{})
+	m.actionCh <- func() {
+		for _, scfg := range cfg.ScrapeConfigs {
+			m.scrapeConfigs[scfg.JobName] = scfg
+		}
+		close(done)
+	}
+	<-done
+	return nil
+}
+
+// Targets returns the targets currently being scraped.
```

Comment:
From this comment I think I'm not understanding what `scrapePool`s actually are. Should `Targets` return all the targets that we're currently scraping at some moment in time, or all targets that the `ScrapeManager` knows about?

Reply:
targets that are currently being scrapped. this function is only used by the web gui - `/targets`